File: packages/sdk/kotlin/client/src/commonMain/kotlin/dev/edgebase/sdk/client/AuthClient.kt
// EdgeBase Kotlin SDK — Authentication client (KMP).
//
// Full auth operations: signUp, signIn, signOut, OAuth, sessions, profile.
//: java.net.URLEncoder → platformUrlEncode() for KMP.

package dev.edgebase.sdk.client

import dev.edgebase.sdk.core.*
import dev.edgebase.sdk.core.generated.GeneratedDbApi

import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow

/**
 * Authentication client for EdgeBase.
 *
 * All HTTP calls delegate to [GeneratedDbApi] where possible.
 * Token management logic is preserved locally.
 *
 * Usage:
 * ```kotlin
 * client.auth.signUp(email = "user@example.com", password = "password")
 * client.auth.signIn(email = "user@example.com", password = "password")
 * ```
 */
class AuthClient(
    private val client: HttpClient,
    private val tokenManager: ClientTokenManager,
    private val core: GeneratedDbApi? = null
) {
    private val _authStateFlow = MutableSharedFlow<Map<String, Any?>?>(
        replay = 1,
        onBufferOverflow = BufferOverflow.DROP_OLDEST
    )

    init {
        tokenManager.setOnAuthStateChange { user ->
            _authStateFlow.tryEmit(user)
        }
    }

    /**
     * Stream of auth state changes.
     */
    val onAuthStateChange: SharedFlow<Map<String, Any?>?> = _authStateFlow.asSharedFlow()

    // MARK: - Authentication

    @Suppress("UNCHECKED_CAST")
    suspend fun signUp(
        email: String,
        password: String,
        data: Map<String, Any>? = null,
        /** Captcha token. */
        captchaToken: String? = null
    ): Map<String, Any?> {
        val body = mutableMapOf<String, Any?>("email" to email, "password" to password)
        if (data != null) body["data"] = data
        val resolved = resolveCaptchaToken(client, "signup", captchaToken)
        if (resolved != null) body["captchaToken"] = resolved
        val result = client.postPublic("/auth/signup", body) as Map<String, Any?>
        handleAuthResponse(result)
        return result
    }

    /**
     * Sign in with email and password.
     * If MFA is enabled, result will contain `mfaRequired=true`, `mfaTicket`, and `factors`.
     * @param captchaToken Captcha token.
     */
    @Suppress("UNCHECKED_CAST")
    suspend fun signIn(
        email: String,
        password: String,
        captchaToken: String? = null
    ): Map<String, Any?> {
        val body = mutableMapOf<String, Any?>("email" to email, "password" to password)
        val resolved = resolveCaptchaToken(client, "signin", captchaToken)
        if (resolved != null) body["captchaToken"] = resolved
        val result = client.postPublic("/auth/signin", body) as Map<String, Any?>
        // If MFA is required, return result without setting tokens
        if (result["mfaRequired"] == true) return result
        handleAuthResponse(result)
        return result
    }

    /** Sign in anonymously. [captchaToken] —. */
    @Suppress("UNCHECKED_CAST")
    suspend fun signInAnonymously(captchaToken: String? = null): Map<String, Any?> {
        val body = mutableMapOf<String, Any?>()
        val resolved = resolveCaptchaToken(client, "anonymous", captchaToken)
        if (resolved != null) body["captchaToken"] = resolved
        val result = client.postPublic("/auth/signin/anonymous", body) as Map<String, Any?>
        handleAuthResponse(result)
        return result
    }

    /**
     * Start OAuth sign-in flow. Returns the OAuth redirect URL.
     * @param captchaToken Captcha token.
     */
    fun signInWithOAuth(provider: String, captchaToken: String? = null): String {
        val base = "${client.baseUrl}/api/auth/oauth/${platformUrlEncode(provider)}"
        if (captchaToken != null) {
            return "$base?captcha_token=${platformUrlEncode(captchaToken)}"
        }
        return base
    }

    /** Send a magic link to the given email. [captchaToken] —. */
    suspend fun signInWithMagicLink(email: String, captchaToken: String? = null) {
        val body = mutableMapOf<String, Any?>("email" to email)
        val resolved = resolveCaptchaToken(client, "magic-link", captchaToken)
        if (resolved != null) body["captchaToken"] = resolved
        client.postPublic("/auth/signin/magic-link", body)
    }

    /** Verify a magic-link token and establish a session. */
    @Suppress("UNCHECKED_CAST")
    suspend fun verifyMagicLink(token: String): Map<String, Any?> {
        val result = client.postPublic("/auth/verify-magic-link", mapOf("token" to token)) as Map<String, Any?>
        handleAuthResponse(result)
        return result
    }

    // ── Phone / SMS Auth ──

    /** Send an SMS verification code to the given phone number. */
    suspend fun signInWithPhone(phone: String, captchaToken: String? = null) {
        val body = mutableMapOf<String, Any?>("phone" to phone)
        val resolved = resolveCaptchaToken(client, "phone", captchaToken)
        if (resolved != null) body["captchaToken"] = resolved
        client.postPublic("/auth/signin/phone", body)
    }

    /** Verify the SMS code and sign in. */
    @Suppress("UNCHECKED_CAST")
    suspend fun verifyPhone(phone: String, code: String): Map<String, Any?> {
        val result = client.postPublic("/auth/verify-phone", mapOf(
            "phone" to phone, "code" to code
        )) as Map<String, Any?>
        handleAuthResponse(result)
        return result
    }

    /** Link current account with a phone number. Sends an SMS code. */
    suspend fun linkWithPhone(phone: String) {
        if (core != null) {
            core.authLinkPhone(mapOf("phone" to phone))
        } else {
            client.post("/auth/link/phone", mapOf("phone" to phone))
        }
    }

    /** Verify phone link code. Completes phone linking for the current account. */
    suspend fun verifyLinkPhone(phone: String, code: String) {
        val body = mapOf<String, Any?>("phone" to phone, "code" to code)
        if (core != null) {
            core.authVerifyLinkPhone(body)
        } else {
            client.post("/auth/verify-link-phone", body)
        }
    }

    @Suppress("UNCHECKED_CAST")
    suspend fun linkWithEmail(email: String, password: String): Map<String, Any?> {
        val body = mapOf<String, Any?>("email" to email, "password" to password)
        val result = if (core != null) {
            core.authLinkEmail(body) as Map<String, Any?>
        } else {
            client.post("/auth/link/email", body) as Map<String, Any?>
        }
        handleAuthResponse(result)
        return result
    }

    /**
     * Link anonymous account to OAuth provider. Returns redirect URL.
     */
    @Suppress("UNCHECKED_CAST")
    suspend fun linkWithOAuth(provider: String, redirectUrl: String = ""): String {
        val body = mapOf<String, Any?>("redirectUrl" to redirectUrl)
        val result = if (core != null) {
            core.oauthLinkStart(provider) as Map<String, Any?>
        } else {
            client.post("/auth/oauth/link/${platformUrlEncode(provider)}", body) as Map<String, Any?>
        }
        return result["redirectUrl"] as? String ?: ""
    }

    suspend fun signOut() {
        // Auto-unregister push token
        try {
            val push = PushClient(client)
            push.unregister()
        } catch (_: Exception) {}

        try {
            val refreshToken = tokenManager.getRefreshToken()
            if (refreshToken != null) {
                if (core != null) {
                    core.authSignout(mapOf("refreshToken" to refreshToken))
                } else {
                    client.post("/auth/signout", mapOf("refreshToken" to refreshToken))
                }
            }
        } catch (_: Exception) {
            // Continue even if server call fails
        }
        tokenManager.clearTokens()
    }

    // MARK: - Session Management

    @Suppress("UNCHECKED_CAST")
    suspend fun listSessions(): List<Map<String, Any?>> {
        val result = if (core != null) {
            core.authGetSessions() as? Map<String, Any?>
        } else {
            client.get("/auth/sessions") as? Map<String, Any?>
        } ?: return emptyList()
        return (result["sessions"] as? List<Map<String, Any?>>) ?: emptyList()
    }

    suspend fun revokeSession(sessionId: String) {
        if (core != null) {
            core.authDeleteSession(sessionId)
        } else {
            client.delete("/auth/sessions/$sessionId")
        }
    }

    @Suppress("UNCHECKED_CAST")
    suspend fun listIdentities(): Map<String, Any?> {
        return if (core != null) {
            core.authGetIdentities() as Map<String, Any?>
        } else {
            client.get("/auth/identities") as Map<String, Any?>
        }
    }

    @Suppress("UNCHECKED_CAST")
    suspend fun unlinkIdentity(identityId: String): Map<String, Any?> {
        return if (core != null) {
            core.authDeleteIdentity(identityId) as Map<String, Any?>
        } else {
            client.delete("/auth/identities/${platformUrlEncode(identityId)}") as Map<String, Any?>
        }
    }

    // MARK: - Profile

    fun currentUser(): Map<String, Any?>? {
        return tokenManager.currentUser()
    }

    /** Fetch the current authenticated user. Delegates to [GeneratedDbApi.authGetMe]. */
    @Suppress("UNCHECKED_CAST")
    suspend fun getMe(): Map<String, Any?> {
        return if (core != null) {
            core.authGetMe() as Map<String, Any?>
        } else {
            client.get("/auth/me") as Map<String, Any?>
        }
    }

    @Suppress("UNCHECKED_CAST")
    suspend fun updateProfile(data: Map<String, Any>): Map<String, Any?> {
        val result = if (core != null) {
            core.authUpdateProfile(data) as Map<String, Any?>
        } else {
            client.patch("/auth/profile", data) as Map<String, Any?>
        }
        handleAuthResponse(result)
        return result
    }

    /** Convenience overload for common profile fields. */
    suspend fun updateProfile(displayName: String? = null, avatarUrl: String? = null): Map<String, Any?> {
        val data = mutableMapOf<String, Any>()
        if (!displayName.isNullOrBlank()) data["displayName"] = displayName
        if (!avatarUrl.isNullOrBlank()) data["avatarUrl"] = avatarUrl
        return updateProfile(data)
    }

    // MARK: - Email Verification / Password Reset

    suspend fun verifyEmail(token: String) {
        client.postPublic("/auth/verify-email", mapOf("token" to token))
    }

    suspend fun verifyEmailChange(token: String) {
        client.postPublic("/auth/verify-email-change", mapOf("token" to token))
    }

    suspend fun requestEmailVerification(redirectUrl: String? = null) {
        val body = mutableMapOf<String, Any?>()
        if (!redirectUrl.isNullOrBlank()) body["redirectUrl"] = redirectUrl
        if (core != null) {
            core.authRequestEmailVerification(body)
        } else {
            client.post("/auth/request-email-verification", body)
        }
    }

    /** @param captchaToken Captcha token. */
    suspend fun requestPasswordReset(email: String, captchaToken: String? = null) {
        val body = mutableMapOf<String, Any?>("email" to email)
        val resolved = resolveCaptchaToken(client, "password-reset", captchaToken)
        if (resolved != null) body["captchaToken"] = resolved
        client.postPublic("/auth/request-password-reset", body)
    }

    suspend fun resetPassword(token: String, newPassword: String) {
        client.postPublic("/auth/reset-password", mapOf(
            "token" to token,
            "newPassword" to newPassword
        ))
    }

    @Suppress("UNCHECKED_CAST")
    suspend fun changePassword(currentPassword: String, newPassword: String): Map<String, Any?> {
        val body = mapOf<String, Any?>("currentPassword" to currentPassword, "newPassword" to newPassword)
        val result = if (core != null) {
            core.authChangePassword(body) as Map<String, Any?>
        } else {
            client.post("/auth/change-password", body) as Map<String, Any?>
        }
        handleAuthResponse(result)
        return result
    }

    @Suppress("UNCHECKED_CAST")
    suspend fun changeEmail(newEmail: String, password: String, redirectUrl: String? = null): Map<String, Any?> {
        require(password.isNotBlank()) { "password is required for changeEmail" }
        val body = mutableMapOf<String, Any?>(
            "newEmail" to newEmail,
            "password" to password,
        )
        if (!redirectUrl.isNullOrBlank()) body["redirectUrl"] = redirectUrl
        return if (core != null) {
            core.authChangeEmail(body) as Map<String, Any?>
        } else {
            client.post("/auth/change-email", body) as Map<String, Any?>
        }
    }

    @Suppress("UNCHECKED_CAST")
    suspend fun refreshToken(): Map<String, Any?> {
        val refreshToken = tokenManager.getRefreshToken()
            ?: throw EdgeBaseError(0, "No refresh token available.")
        val result = client.postPublic(
            "/auth/refresh",
            mapOf("refreshToken" to refreshToken)
        ) as Map<String, Any?>
        handleAuthResponse(result)
        return result
    }

    // MARK: - Passkeys / WebAuthn REST layer

    @Suppress("UNCHECKED_CAST")
    suspend fun passkeysRegisterOptions(): Map<String, Any?> {
        return if (core != null) {
            core.authPasskeysRegisterOptions() as Map<String, Any?>
        } else {
            client.post("/auth/passkeys/register-options", emptyMap<String, Any>()) as Map<String, Any?>
        }
    }

    @Suppress("UNCHECKED_CAST")
    suspend fun passkeysRegister(response: Any?): Map<String, Any?> {
        val body = mapOf("response" to response)
        return if (core != null) {
            core.authPasskeysRegister(body) as Map<String, Any?>
        } else {
            client.post("/auth/passkeys/register", body) as Map<String, Any?>
        }
    }

    @Suppress("UNCHECKED_CAST")
    suspend fun passkeysAuthOptions(email: String? = null): Map<String, Any?> {
        val body = if (email.isNullOrBlank()) emptyMap() else mapOf("email" to email)
        return if (core != null) {
            core.authPasskeysAuthOptions(body) as Map<String, Any?>
        } else {
            client.postPublic("/auth/passkeys/auth-options", body) as Map<String, Any?>
        }
    }

    @Suppress("UNCHECKED_CAST")
    suspend fun passkeysAuthenticate(response: Any?): Map<String, Any?> {
        val body = mapOf("response" to response)
        val result = if (core != null) {
            core.authPasskeysAuthenticate(body) as Map<String, Any?>
        } else {
            client.postPublic("/auth/passkeys/authenticate", body) as Map<String, Any?>
        }
        handleAuthResponse(result)
        return result
    }

    @Suppress("UNCHECKED_CAST")
    suspend fun passkeysList(): Map<String, Any?> {
        return if (core != null) {
            core.authPasskeysList() as Map<String, Any?>
        } else {
            client.get("/auth/passkeys") as Map<String, Any?>
        }
    }

    @Suppress("UNCHECKED_CAST")
    suspend fun passkeysDelete(credentialId: String): Map<String, Any?> {
        return if (core != null) {
            core.authPasskeysDelete(credentialId) as Map<String, Any?>
        } else {
            client.delete("/auth/passkeys/$credentialId") as Map<String, Any?>
        }
    }

    // MARK: - MFA / TOTP

    /** MFA sub-namespace for TOTP enrollment, verification, and management. */
    val mfa = MfaClient()

    inner class MfaClient {
        /** Enroll TOTP — returns factorId, secret, qrCodeUri, and recoveryCodes. */
        @Suppress("UNCHECKED_CAST")
        suspend fun enrollTotp(): Map<String, Any?> {
            return if (core != null) {
                core.authMfaTotpEnroll() as Map<String, Any?>
            } else {
                client.post("/auth/mfa/totp/enroll", emptyMap<String, Any>()) as Map<String, Any?>
            }
        }

        /** Verify TOTP enrollment with factorId and a TOTP code. */
        suspend fun verifyTotpEnrollment(factorId: String, code: String) {
            if (core != null) {
                core.authMfaTotpVerify(mapOf("factorId" to factorId, "code" to code))
            } else {
                client.post("/auth/mfa/totp/verify", mapOf("factorId" to factorId, "code" to code))
            }
        }

        /**
         * Verify TOTP code during MFA challenge (after signIn returns mfaRequired).
         * Note: Uses postPublic (no auth) since user is mid-signin.
         */
        @Suppress("UNCHECKED_CAST")
        suspend fun verifyTotp(mfaTicket: String, code: String): Map<String, Any?> {
            // Public endpoint — user has no token yet during MFA challenge
            val result = client.postPublic("/auth/mfa/verify", mapOf(
                "mfaTicket" to mfaTicket, "code" to code
            )) as Map<String, Any?>
            handleAuthResponse(result)
            return result
        }

        /**
         * Use a recovery code during MFA challenge.
         * Note: Uses postPublic (no auth) since user is mid-signin.
         */
        @Suppress("UNCHECKED_CAST")
        suspend fun useRecoveryCode(mfaTicket: String, recoveryCode: String): Map<String, Any?> {
            // Public endpoint — user has no token yet during MFA challenge
            val result = client.postPublic("/auth/mfa/recovery", mapOf(
                "mfaTicket" to mfaTicket, "recoveryCode" to recoveryCode
            )) as Map<String, Any?>
            handleAuthResponse(result)
            return result
        }

        /** Disable TOTP for the current user. Requires password or TOTP code. */
        suspend fun disableTotp(password: String? = null, code: String? = null) {
            val body = mutableMapOf<String, Any?>()
            if (password != null) body["password"] = password
            if (code != null) body["code"] = code
            if (core != null) {
                core.authMfaTotpDelete(body)
            } else {
                client.delete("/auth/mfa/totp", body)
            }
        }

        /** List enrolled MFA factors for the current user. */
        @Suppress("UNCHECKED_CAST")
        suspend fun listFactors(): List<Map<String, Any?>> {
            val result = if (core != null) {
                core.authMfaFactors() as? Map<String, Any?>
            } else {
                client.get("/auth/mfa/factors") as? Map<String, Any?>
            } ?: return emptyList()
            return (result["factors"] as? List<Map<String, Any?>>) ?: emptyList()
        }
    }

    // MARK: - Internal

    private suspend fun handleAuthResponse(result: Map<String, Any?>) {
        val accessToken = result["accessToken"] as? String
        val refreshToken = result["refreshToken"] as? String
        if (accessToken != null && refreshToken != null) {
            tokenManager.setTokens(TokenPair(accessToken, refreshToken))
        }
    }

    suspend fun signInWithEmailOtp(email: String) {
        val body = mapOf<String, Any?>("email" to email)
        if (core != null) {
            core.authSigninEmailOtp(body)
        } else {
            client.post("/auth/signin/email-otp", body)
        }
    }
}

File: packages/sdk/swift/packages/ios/Sources/AuthClient.swift
import EdgeBaseCore
// Auth Client — user authentication.
//
// Mirrors Dart SDK AuthClient:
// signUp / signIn / signOut / signInAnonymously
// OAuth / linkWithEmail / linkWithOAuth
// sessions / profile / email verify / password reset
// onAuthStateChange via AsyncStream

import Foundation

/// Authentication client for user operations.
public final class AuthClient: @unchecked Sendable {
    public struct PasskeysAuthOptions {
        public let email: String?

        public init(email: String? = nil) {
            self.email = email
        }

        var body: [String: Any] {
            var result: [String: Any] = [:]
            if let email, !email.isEmpty {
                result["email"] = email
            }
            return result
        }
    }

    private let client: HttpClient
    private let tokenManager: TokenManager
    private let core: GeneratedDbApi

    public init(client: HttpClient, tokenManager: TokenManager, core: GeneratedDbApi? = nil) {
        self.client = client
        self.tokenManager = tokenManager
        self.core = core ?? GeneratedDbApi(http: client)
    }

    // MARK: - Core Auth

    /// Sign up with email and password.
    /// - Parameter captchaToken: Captcha token.
    @discardableResult
    public func signUp(email: String, password: String, userData: [String: Any]? = nil, captchaToken: String? = nil) async throws -> [String: Any] {
        var body: [String: Any] = ["email": email, "password": password]
        if let userData = userData {
            body["data"] = userData
        }
        let resolved = await TurnstileProvider.resolveCaptchaToken(core: core, baseUrl: await client.baseUrl, action: "signup", manualToken: captchaToken)
        if let resolved { body["captchaToken"] = resolved }
        let result = try expectObject(await client.postPublic("/auth/signup", body))
        if let tokens = parseTokens(result) {
            await tokenManager.setTokens(tokens)
        }
        return result
    }

    /// Sign in with email and password.
    /// Returns dictionary that may contain `mfaRequired: true` if MFA is enabled.
    /// - Parameter captchaToken: Captcha token.
    public func signIn(email: String, password: String, captchaToken: String? = nil) async throws -> [String: Any] {
        var body: [String: Any] = ["email": email, "password": password]
        let resolved = await TurnstileProvider.resolveCaptchaToken(core: core, baseUrl: await client.baseUrl, action: "signin", manualToken: captchaToken)
        if let resolved { body["captchaToken"] = resolved }
        let result = try expectObject(await client.postPublic("/auth/signin", body))
        // If MFA is required, return the result without setting tokens
        if result["mfaRequired"] as? Bool == true {
            return result
        }
        if let tokens = parseTokens(result) {
            await tokenManager.setTokens(tokens)
        }
        return result
    }

    /// Sign out.
    public func signOut() async {
        // Auto-unregister push token
        do {
            let push = PushClient(client)
            try await push.unregister()
        } catch {}

        do {
            let refreshToken = await tokenManager.getRefreshToken()
            if let refreshToken = refreshToken {
                try await client.post("/auth/signout", ["refreshToken": refreshToken])
            }
        } catch {
            // Continue even if server call fails
        }
        await tokenManager.clearTokens()
    }

    /// Sign in anonymously.
    /// - Parameter captchaToken: Captcha token.
    public func signInAnonymously(captchaToken: String? = nil) async throws -> [String: Any] {
        var body: [String: Any] = [:]
        let resolved = await TurnstileProvider.resolveCaptchaToken(core: core, baseUrl: await client.baseUrl, action: "anonymous", manualToken: captchaToken)
        if let resolved { body["captchaToken"] = resolved }
        let result = try expectObject(await client.postPublic("/auth/signin/anonymous", body))
        if let tokens = parseTokens(result) {
            await tokenManager.setTokens(tokens)
        }
        return result
    }

    // MARK: - OAuth

    /// Start OAuth flow — returns the OAuth redirect URL.
    /// Open this URL in a browser (SFSafariViewController, etc.) to initiate.
    /// - Parameter captchaToken: Captcha token.
    public func signInWithOAuth(provider: String, captchaToken: String? = nil) async -> String {
        let encoded = provider.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? provider
        var url = await client.apiUrl("/auth/oauth/\(encoded)")
        if let captchaToken = captchaToken,
           let encodedToken = captchaToken.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) {
            url += "?captcha_token=\(encodedToken)"
        }
        return url
    }

    // MARK: - Magic Link

    /// Request a magic link sign-in email.
    /// - Parameter captchaToken: Captcha token.
    public func signInWithMagicLink(email: String, captchaToken: String? = nil) async throws {
        var body: [String: Any] = ["email": email]
        let resolved = await TurnstileProvider.resolveCaptchaToken(core: core, baseUrl: await client.baseUrl, action: "magic-link", manualToken: captchaToken)
        if let resolved { body["captchaToken"] = resolved }
        _ = try await client.postPublic("/auth/signin/magic-link", body)
    }

    /// Verify a magic link token and sign in.
    public func verifyMagicLink(token: String) async throws -> [String: Any] {
        let result = try expectObject(await client.postPublic("/auth/verify-magic-link", ["token": token]))
        if let tokens = parseTokens(result) {
            await tokenManager.setTokens(tokens)
        }
        return result
    }

    /// Verify an email OTP code and sign in.
    public func verifyEmailOtp(email: String, code: String) async throws -> [String: Any] {
        let result = try expectObject(await client.postPublic("/auth/verify-email-otp", [
            "email": email,
            "code": code,
        ]))
        if let tokens = parseTokens(result) {
            await tokenManager.setTokens(tokens)
        }
        return result
    }

    // MARK: - Phone / SMS Auth

    /// Send an SMS verification code to the given phone number.
    /// - Parameter captchaToken: Captcha token.
    public func signInWithPhone(phone: String, captchaToken: String? = nil) async throws {
        var body: [String: Any] = ["phone": phone]
        let resolved = await TurnstileProvider.resolveCaptchaToken(core: core, baseUrl: await client.baseUrl, action: "phone", manualToken: captchaToken)
        if let resolved { body["captchaToken"] = resolved }
        _ = try await client.postPublic("/auth/signin/phone", body)
    }

    /// Verify the SMS code and sign in.
    public func verifyPhone(phone: String, code: String) async throws -> [String: Any] {
        let result = try expectObject(await client.postPublic("/auth/verify-phone", [
            "phone": phone, "code": code
        ]))
        if let tokens = parseTokens(result) {
            await tokenManager.setTokens(tokens)
        }
        return result
    }

    /// Link current account with a phone number. Sends an SMS code.
    public func linkWithPhone(phone: String) async throws {
        _ = try await client.post("/auth/link/phone", ["phone": phone])
    }

    /// Verify phone link code. Completes phone linking for the current account.
    public func verifyLinkPhone(phone: String, code: String) async throws {
        _ = try await client.post("/auth/verify-link-phone", [
            "phone": phone, "code": code
        ])
    }

    /// Link current account with email.
    public func linkWithEmail(email: String, password: String) async throws -> [String: Any] {
        let body: [String: Any] = ["email": email, "password": password]
        let result = try expectObject(await client.post("/auth/link/email", body))
        if let tokens = parseTokens(result) {
            await tokenManager.setTokens(tokens)
        }
        return result
    }

    /// Link current account with OAuth provider. Returns redirect URL.
    public func linkWithOAuth(provider: String, redirectUrl: String = "") async throws -> String {
        let encoded = provider.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? provider
        let body: [String: Any] = ["redirectUrl": redirectUrl]
        let result = try expectObject(await client.post("/auth/oauth/link/\(encoded)", body))
        return (result["redirectUrl"] as? String) ?? ""
    }

    // MARK: - Sessions

    /// List active sessions.
    public func listSessions() async throws -> [[String: Any]] {
        let result = try await client.get("/auth/sessions")
        if let sessions = result as? [[String: Any]] {
            return sessions
        }
        if let dict = result as? [String: Any], let sessions = dict["sessions"] as? [[String: Any]] {
            return sessions
        }
        return []
    }

    /// Revoke a session by ID.
    public func revokeSession(_ sessionId: String) async throws {
        try await client.delete("/auth/sessions/\(sessionId)")
    }

    /// Compatibility overload with an external label.
    public func revokeSession(sessionId: String) async throws {
        try await revokeSession(sessionId)
    }

    /// List linked sign-in identities for the current user.
    public func listIdentities() async throws -> [String: Any] {
        return try expectObject(await core.authGetIdentities())
    }

    /// Unlink a linked OAuth identity by its identity ID.
    public func unlinkIdentity(_ identityId: String) async throws -> [String: Any] {
        let encoded = identityId.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? identityId
        return try expectObject(await client.delete("/auth/identities/\(encoded)"))
    }

    // MARK: - Profile

    /// Update current user profile.
    public func updateProfile(_ data: [String: Any]) async throws -> [String: Any] {
        let result = try expectObject(await client.patch("/auth/profile", data))
        if let tokens = parseTokens(result) {
            await tokenManager.setTokens(tokens)
        }
        return result
    }

    /// Convenience overload for common profile fields.
    public func updateProfile(displayName: String? = nil, avatarUrl: String? = nil) async throws -> [String: Any] {
        var data: [String: Any] = [:]
        if let displayName, !displayName.isEmpty {
            data["displayName"] = displayName
        }
        if let avatarUrl, !avatarUrl.isEmpty {
            data["avatarUrl"] = avatarUrl
        }
        return try await updateProfile(data)
    }

    /// Get current user from token.
    public func currentUser() async -> [String: Any]? {
        return await tokenManager.currentUser()
    }

    // MARK: - Email Verification & Password Reset

    /// Request email verification.
    public func verifyEmail(token: String) async throws -> [String: Any] {
        return try expectObject(await client.postPublic("/auth/verify-email", ["token": token]))
    }

    /// Request a verification email for the current authenticated user.
    public func requestEmailVerification(redirectUrl: String? = nil) async throws -> [String: Any] {
        var body: [String: Any] = [:]
        if let redirectUrl, !redirectUrl.isEmpty {
            body["redirectUrl"] = redirectUrl
        }
        return try expectObject(await client.post("/auth/request-email-verification", body))
    }

    /// Verify a pending email change using the emailed token.
    public func verifyEmailChange(token: String) async throws -> [String: Any] {
        return try expectObject(await client.postPublic("/auth/verify-email-change", ["token": token]))
    }

    /// Request password reset email.
    /// - Parameter captchaToken: Captcha token.
    public func requestPasswordReset(email: String, captchaToken: String? = nil) async throws -> [String: Any] {
        var body: [String: Any] = ["email": email]
        let resolved = await TurnstileProvider.resolveCaptchaToken(core: core, baseUrl: await client.baseUrl, action: "password-reset", manualToken: captchaToken)
        if let resolved { body["captchaToken"] = resolved }
        return try expectObject(await client.postPublic("/auth/request-password-reset", body))
    }

    /// Reset password with token.
    public func resetPassword(token: String, newPassword: String) async throws -> [String: Any] {
        let body: [String: Any] = ["token": token, "newPassword": newPassword]
        return try expectObject(await client.postPublic("/auth/reset-password", body))
    }

    /// Change password for authenticated user.
    public func changePassword(currentPassword: String, newPassword: String) async throws -> [String: Any] {
        let body: [String: Any] = ["currentPassword": currentPassword, "newPassword": newPassword]
        let result = try expectObject(await client.post("/auth/change-password", body))
        if let tokens = parseTokens(result) {
            await tokenManager.setTokens(tokens)
        }
        return result
    }

    /// Request an email change for the current user.
    public func changeEmail(newEmail: String, password: String, redirectUrl: String? = nil) async throws -> [String: Any] {
        var body: [String: Any] = [
            "newEmail": newEmail,
            "password": password,
        ]
        if let redirectUrl, !redirectUrl.isEmpty {
            body["redirectUrl"] = redirectUrl
        }
        return try expectObject(await client.post("/auth/change-email", body))
    }

    /// Refresh the current session using the stored refresh token.
    public func refreshToken() async throws -> [String: Any] {
        guard let refreshToken = await tokenManager.getRefreshToken(),
              !refreshToken.isEmpty else {
            throw EdgeBaseError(statusCode: 0, message: "No refresh token available.")
        }
        let result = try expectObject(await client.postPublic("/auth/refresh", ["refreshToken": refreshToken]))
        if let tokens = parseTokens(result) {
            await tokenManager.setTokens(tokens)
        }
        return result
    }

    // MARK: - Auth State Change

    /// Stream of auth state changes using AsyncStream.
    /// Usage: `for await user in auth.onAuthStateChange { ... }`
    public func onAuthStateChange() -> AsyncStream<[String: Any]?> {
        AsyncStream { continuation in
            Task {
                await tokenManager.onAuthStateChange { user in
                    continuation.yield(user)
                }
            }
        }
    }

    // MARK: - Passkeys / WebAuthn REST layer

    public func passkeysRegisterOptions() async throws -> [String: Any] {
        try expectObject(await core.authPasskeysRegisterOptions())
    }

    public func passkeysRegister(response: [String: Any]) async throws -> [String: Any] {
        try expectObject(await core.authPasskeysRegister(["response": response]))
    }

    public func passkeysAuthOptions(email: String? = nil) async throws -> [String: Any] {
        try await passkeysAuthOptions(PasskeysAuthOptions(email: email))
    }

    public func passkeysAuthOptions(_ options: PasskeysAuthOptions) async throws -> [String: Any] {
        try expectObject(await core.authPasskeysAuthOptions(options.body))
    }

    public func passkeysAuthenticate(response: [String: Any]) async throws -> [String: Any] {
        let result = try expectObject(await core.authPasskeysAuthenticate(["response": response]))
        if let tokens = parseTokens(result) {
            await tokenManager.setTokens(tokens)
        }
        return result
    }

    public func passkeysList() async throws -> [String: Any] {
        try expectObject(await core.authPasskeysList())
    }

    public func passkeysDelete(credentialId: String) async throws -> [String: Any] {
        try expectObject(await core.authPasskeysDelete(credentialId))
    }

    // MARK: - MFA / TOTP

    /// Enroll TOTP — returns factorId, secret, qrCodeUri, and recoveryCodes.
    public func enrollTotp() async throws -> [String: Any] {
        return try expectObject(await client.post("/auth/mfa/totp/enroll", [:]))
    }

    /// Compatibility helper matching other SDKs.
    /// Verify TOTP enrollment with factorId and a TOTP code.
    public func verifyTotpEnrollment(factorId: String, code: String) async throws {
        _ = try await client.post("/auth/mfa/totp/verify", ["factorId": factorId, "code": code])
    }

    /// Verify TOTP code during MFA challenge (after signIn returns mfaRequired).
    public func verifyTotp(mfaTicket: String, code: String) async throws -> [String: Any] {
        let result = try expectObject(await client.postPublic("/auth/mfa/verify", [
            "mfaTicket": mfaTicket, "code": code
        ]))
        if let tokens = parseTokens(result) {
            await tokenManager.setTokens(tokens)
        }
        return result
    }

    /// Use a recovery code during MFA challenge.
    public func useRecoveryCode(mfaTicket: String, recoveryCode: String) async throws -> [String: Any] {
        let result = try expectObject(await client.postPublic("/auth/mfa/recovery", [
            "mfaTicket": mfaTicket, "recoveryCode": recoveryCode
        ]))
        if let tokens = parseTokens(result) {
            await tokenManager.setTokens(tokens)
        }
        return result
    }

    /// Disable TOTP for the current user. Requires password or TOTP code.
    public func disableTotp(password: String? = nil, code: String? = nil) async throws {
        var body: [String: Any] = [:]
        if let password = password { body["password"] = password }
        if let code = code { body["code"] = code }
        try await client.delete("/auth/mfa/totp", body)
    }

    /// List enrolled MFA factors for the current user.
    public func listFactors() async throws -> [[String: Any]] {
        let result = try await client.get("/auth/mfa/factors")
        if let dict = result as? [String: Any], let factors = dict["factors"] as? [[String: Any]] {
            return factors
        }
        return []
    }

    /// Request an OTP code for email sign-in.
    public func signInWithEmailOtp(email: String) async throws -> [String: Any] {
        try expectObject(await client.post("/auth/signin/email-otp", ["email": email]))
    }

    // MARK: - Internal

    /// Safely cast a decoded JSON response to an object, throwing on unexpected shapes
    /// (e.g. a bare array, string, or null) instead of crashing on a force-cast.
    private func expectObject(_ value: Any) throws -> [String: Any] {
        guard let object = value as? [String: Any] else {
            throw EdgeBaseError(statusCode: 0, message: "Unexpected auth response shape: expected a JSON object.")
        }
        return object
    }

    private func parseTokens(_ result: [String: Any]) -> TokenPair? {
        guard let accessToken = result["accessToken"] as? String,
              let refreshToken = result["refreshToken"] as? String else {
            return nil
        }
        return TokenPair(accessToken: accessToken, refreshToken: refreshToken)
    }
}
